File: dlang/plugin-api/src/main/java/io/github/intellij/dlanguage/stubs/types/DStubElementType.java
package io.github.intellij.dlanguage.stubs.types;

import com.intellij.lang.ASTNode;
import com.intellij.psi.stubs.IStubElementType;
import com.intellij.psi.stubs.IndexSink;
import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubOutputStream;
import io.github.intellij.dlanguage.DLanguage;
import io.github.intellij.dlanguage.psi.interfaces.DCompositeElement;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;


public abstract class DStubElementType<S extends StubElement<T>, T extends DCompositeElement> extends IStubElementType<S, T> {
    public DStubElementType(final String debugName) {
        super(debugName, DLanguage.INSTANCE);
    }

    @Override
    public void indexStub(@NotNull final S stub, @NotNull final IndexSink sink) {

    }

    @Override
    public void serialize(@NotNull final S stub, @NotNull final StubOutputStream dataStream) throws IOException {

    }

    @NotNull
    @Override
    public String getExternalId() {
        return "d." + super.toString();
    }

    @Override
    public boolean shouldCreateStub(final ASTNode node) {
        return true;
    }
}

File: dlang/psi-impl/src/main/kotlin/io/github/intellij/dlanguage/stubs/types/DStubElementType.kt
package io.github.intellij.dlanguage.stubs.types

import com.intellij.psi.PsiElement
import com.intellij.psi.stubs.IStubElementType
import com.intellij.psi.stubs.IndexSink
import com.intellij.psi.stubs.StubElement
import com.intellij.psi.stubs.StubOutputStream
import io.github.intellij.dlanguage.DLanguage
import java.io.IOException

abstract class DStubElementType<S : StubElement<T>, T : PsiElement>(debugName: String) :
    IStubElementType<S, T>(debugName, DLanguage) {
    override fun indexStub(stub: S, sink: IndexSink) {
    }

    @Throws(IOException::class)
    override fun serialize(stub: S, dataStream: StubOutputStream) {
    }

    override fun getExternalId(): String {
        return "d." + super.toString()
    }
}

